pkg/streamingpromql/operators/aggregations/topkbottomk/instant_query.go

```diff
@@ -0,0 +1,320 @@
+	// (ie. larger for topk / smaller for bottomk)
+
+	currentWorstValue := g.series[0].value
+
+	if math.IsNaN(value) {
+		// A NaN is never better than any existing value.
+		return false
+	} else if t.IsTopK && value <= currentWorstValue && !math.IsNaN(currentWorstValue) {
+		// Value is not larger than the nth biggest value we've already seen. Continue.
+		return false
+	} else if !t.IsTopK && value >= currentWorstValue && !math.IsNaN(currentWorstValue) {
+		// Value is not smaller than the nth smallest value we've already seen. Continue.
+		return false
+	}
+
+	g.series[0].metadata = metadata
+	g.series[0].value = value
+
+	if t.limit != 1 {
+		// We only need to bother to fix the heap if there's more than one element.
+		// This optimises for the common case of topk(1, xxx) or bottomk(1, xxx).
+		t.heap.Reset(g)
+		heap.Fix(t.heap, 0)
+	}
+
+	if math.IsNaN(currentWorstValue) {
+		g.nanCount--
+	}
+
+	return false
+}
+
+func (t *InstantQuery) NextSeries(_ context.Context) (types.InstantVectorSeriesData, error) {
+	if t.nextSeriesIndex >= len(t.values) {
+		return types.InstantVectorSeriesData{}, types.EOS
+	}
+
+	data := types.InstantVectorSeriesData{}
+
+	var err error
+	data.Floats, err = types.FPointSlicePool.Get(1, t.MemoryConsumptionTracker)
+	if err != nil {
+		return types.InstantVectorSeriesData{}, err
+	}
+
+	data.Floats = append(data.Floats, promql.FPoint{
+		T: t.TimeRange.StartT,
+		F: t.values[t.nextSeriesIndex],
+	})
+
+	t.nextSeriesIndex++
+
+	return data, nil
+}
+
+func (t *InstantQuery) ExpressionPosition() posrange.PositionRange {
+	return t.expressionPosition
+}
+
+func (t *InstantQuery) Close() {
+	t.Inner.Close()
+	t.Param.Close()
+
+	types.Float64SlicePool.Put(t.values, t.MemoryConsumptionTracker)
+}
+
+type instantQueryGroup struct {
+	seriesCount int // Total number of series that contribute to this group.
+	nanCount    int // Total number of output series (ie. in the heap below) whose value is NaN.
+
+	series []instantQuerySeries // Min-/max-heap for the current 'best' values we've seen (highest for topk / lowest for bottomk)
+}
+
+type instantQuerySeries struct {
+	metadata types.SeriesMetadata
+	value    float64
+}
+
+type instantQueryHeap struct {
+	group *instantQueryGroup
+	less  func(i, j float64) bool
+}
+
+func (h *instantQueryHeap) Reset(group *instantQueryGroup) {
+	h.group = group
+}
+
+func (h *instantQueryHeap) Len() int {
+	return len(h.group.series)
+}
+
+func (h *instantQueryHeap) Less(i, j int) bool {
+	iValue := h.group.series[i].value
+	jValue := h.group.series[j].value
+
+	return h.less(iValue, jValue)
+}
+
+func (h *instantQueryHeap) Swap(i, j int) {
+	h.group.series[i], h.group.series[j] = h.group.series[j], h.group.series[i]
+}
+
+func (h *instantQueryHeap) Push(x any) {
+	h.group.series = append(h.group.series, x.(instantQuerySeries))
+}
+
+func (h *instantQueryHeap) Pop() any {
+	i := len(h.group.series) - 1
+	v := h.group.series[i]
+	h.group.series = h.group.series[:i]
+	return v
+}
```

Comment:
We should assert `instantQueryHeap` implements `heap.Interface`

Reply:
This is covered by usage - this file will fail to compile if `*instantQueryHeap` doesn't implement `heap.Interface` because `t.heap` is passed to methods like `heap.Push` and `heap.Fix` in `accumulateValue`.

Reply:
I personally prefer to be explicit for clarity where as it puts the expected interface next to the definition, but I'm not blocked on it.